src/App.js

```diff
@@ -1,79 +1,154 @@
+  };
 
+  // Adding data functions for firebase, electron adn Mturk
+  const defaultFunction = (data) => {};
+  const firebaseUpdateFunction = (data) => {
+    addToFirebase(data);
+  };
+  const desktopUpdateFunction = (data) => {
+    ipcRenderer.send("data", data);
+  };
+  const psiturkUpdateFunction = (data) => {
+    psiturk.recordTrialData(data);
+  };
 
-const isElectron = !MTURK
-let ipcRenderer = false;
-let psiturk = false
-if (isElectron) {
-  const electron = window.require('electron');
-  ipcRenderer  = electron.ipcRenderer;
-} else {
-  /* eslint-disable */
-  window.lodash = _.noConflict()
-  psiturk = new PsiTurk(getTurkUniqueId(), '/complete')
-  /* eslint-enable */
-}
+  // On finish functions for electron, Mturk
+  const desktopFinishFunction = () => {
+    ipcRenderer.send("end", "true");
+  };
+  const psiturkFinishFunction = () => {
+    const completePsiturk = async () => {
+      psiturk.saveData();
+      await sleep(5000);
+      psiturk.completeHIT();
+    };
+    completePsiturk();
+  };
+
+  // Login logic
+  useEffect(() => {
+    // For testing and debugging purposes
+    console.log("MTURK? :" + MTURK);
+    console.log("ELECTRON? :" + IS_ELECTRON);
+    console.log("AT_HOME? :" + AT_HOME);
+    console.log("FIREBASE? :" + FIREBASE);
 
-class App extends React.Component {
-  render() {
-    console.log("Outside Turk:", jsPsych.turk.turkInfo().outsideTurk)
-    console.log("Turk:", MTURK)
-    jsPsych.plugins['rt-categorize-html'] = rt_categorize_html();
-    
-    return (
-      <div className="App">
-        <Experiment settings={{
-          timeline: tl,
-          on_data_update: (data) => {
-            //firebase 
-            if(FIREBASE){
-              if (data.trial_index === 1) {
-                
-                console.log(data.patient_id);// delete when done
-  
-                createFirebaseDocument(data.patient_id);
-                addToFirebase(data);
-              }
-              if (data.trial_index > 1) {
-                addToFirebase(data);
-              }
-            }
-            else if ( ipcRenderer ) {
-              ipcRenderer.send('data', data)
-            }
-            else if (psiturk) {
-                psiturk.recordTrialData(data)
-            }
-          },
-          on_finish: (data) => {
-            if ( ipcRenderer ) {
-              ipcRenderer.send('end', 'true')
-            }
-            else if (psiturk) {
-              const completePsiturk = async () => {
-                psiturk.saveData()
-                await sleep(5000)
-                psiturk.completeHIT()
-              }
-              completePsiturk()
-            }
-          },
-        }}
+    // If on desktop
+    if (IS_ELECTRON) {
+      const electron = window.require("electron");
+      const renderer = electron.ipcRenderer;
+      setRenderer(renderer);
+      // If at home, fill in fields based on environment variables
+      if (AT_HOME) {
```

Comment:
I don't think this should be specific to at_home, as in the clinic the study ID could definitely be set this way as well.

Reply:
then you wouldn't need the else on 78